File: part03/solutions/scala/src/main/scala/com/waitingforcode/DataGenerator.scala
package com.waitingforcode

import org.apache.spark.sql.SparkSession

import java.sql.Timestamp
import java.time.{ZoneId, ZonedDateTime}
import java.util.UUID
import java.util.concurrent.{ThreadLocalRandom, TimeUnit}
import scala.sys.process.Process

object DataGenerator {

  def main(args: Array[String]): Unit = {
    Seq(InputDataTopicName, DedupedDataTopicName).foreach(topicToRecreate => {
      println(s"== Deleting already existing topic ${topicToRecreate} ==")
      val deleteTopicResult =
        Process(s"docker exec docker_kafka_1 kafka-topics.sh --bootstrap-server localhost:9092 --topic ${topicToRecreate} --delete").run()
      deleteTopicResult.exitValue()
      println(s"== Creating ${topicToRecreate} ==")
      val createTopicResult =
        Process(s"docker exec docker_kafka_1 kafka-topics.sh --bootstrap-server localhost:9092 --topic ${topicToRecreate} --create --partitions 2").run()
      createTopicResult.exitValue()
    })
    val sparkSession = SparkSession.builder().master("local[*]").getOrCreate()
    import sparkSession.implicits._

    while (true) {
      println(">>> Generating records")
      def generateVisit(visitId: Int): Visit = {
        Visit(event_id = UUID.randomUUID().toString, visit_id = visitId,
          event_time = Timestamp.from(ZonedDateTime.now(ZoneId.of("UTC")).toInstant))
      }

      val maybeDuplicatedVisit2 = generateVisit(2)
      val maybeDuplicatedVisit3 = generateVisit(3)
      val maybeDuplicatedVisit4 = generateVisit(4)
      val extraVisits = if (ThreadLocalRandom.current().nextBoolean()) {
        Seq(maybeDuplicatedVisit2, maybeDuplicatedVisit3, maybeDuplicatedVisit4)
      } else {
        Seq.empty
      }
      val visits = (Seq(
        generateVisit(1), generateVisit(1), generateVisit(1),
        generateVisit(2), generateVisit(2), maybeDuplicatedVisit2,
        generateVisit(3), generateVisit(3), maybeDuplicatedVisit3,
        generateVisit(4), maybeDuplicatedVisit4
      ) ++ extraVisits)

      visits.toDS().selectExpr("TO_JSON(STRUCT(*)) AS value")
        .write
        .format("kafka")
        .options(
          Map(
            "kafka.bootstrap.servers" -> "localhost:9094",
            "topic" -> InputDataTopicName,
          )
        )
        .save()

      Thread.sleep(TimeUnit.SECONDS.toMillis(5))
    }
  }

}

File: part03/solutions/python/data_generator.py
import json
import os
import random
import uuid
from datetime import datetime
import time
from typing import Dict

from pyspark.sql import SparkSession

from config import kafka_input_topic, kafka_deduped_topic

for topic in [kafka_input_topic(), kafka_deduped_topic()]:
    print('>>> Recreating topics...')
    os.system(
        f'docker exec docker_kafka_1 kafka-topics.sh --bootstrap-server localhost:9092 --topic {topic} --delete')
    os.system(
        f'docker exec docker_kafka_1 kafka-topics.sh --bootstrap-server localhost:9092 --topic {topic} --create --partitions 2')

spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.5.0') \
    .getOrCreate()


def visit(visit_id: int) -> Dict[str, str]:
    return {'value': json.dumps({'event_id': str(uuid.uuid4()), 'visit_id': visit_id,
                                 'event_time': datetime.now().isoformat(),
                                 'page': random.choice(
                                     ["page1", "page2", "page3", "page4", "page5"])})}


while True:
    print('Generating data')
    maybe_duplicated_visit_2 = visit(2)
    maybe_duplicated_visit_3 = visit(3)
    maybe_duplicated_visit_4 = visit(4)
    duplicates = []

    if bool(random.getrandbits(1)):
        duplicates.append(maybe_duplicated_visit_2)
        duplicates.append(maybe_duplicated_visit_3)
        duplicates.append(maybe_duplicated_visit_4)

    dataset = [visit(1), visit(1), visit(1),
               visit(2), visit(2), maybe_duplicated_visit_2,
               visit(3), visit(3), maybe_duplicated_visit_3,
               visit(4), maybe_duplicated_visit_4] + duplicates

    dataframe_to_write = spark.createDataFrame(dataset, "value STRING") \
        .selectExpr("value")
    dataframe_to_write.write.format('kafka').option('kafka.bootstrap.servers', 'localhost:9094') \
        .option('topic', kafka_input_topic()).save()

    time.sleep(5)